server/aesdsocket.c

```diff
@@ -1,331 +1,580 @@
+
+struct thread_t
+{
+    pthread_t thread_id;
+    int client_socket_t;
+    int server_socket_t;
+    int logfile_fd;
+    pthread_mutex_t *mutex;
+    bool complete;
+    int *file_size;
+    struct sockaddr_storage client_address;
+
+    //singly linked list node 
+    SLIST_ENTRY(thread_t) node;
+};
+
+//Global variables
+bool graceful_exit_handle = false;
+int file_fd = 0;
+int file_size = 0;
+
+//Mutex initialization
+pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+
+
+//Signal handler for SIGINT and SIGTERM
+void signal_handler(int sig)
+{
+    if(sig == SIGINT || sig == SIGTERM)
+    {
+       syslog(LOG_INFO, "Caught SIGINT or SIGTERM. Exiting...");    
+       graceful_exit_handle = true;
+    }
+}
+
+//Handles printing timestamp every 10 seconds
+void timestamp_handler(int sig_id)
+{
+    if(sig_id == SIGALRM)
+    {
+        //printing timestamp after 10 seconds, when SIGALRM is raised
+        int unlock_ret = pthread_mutex_lock(&mutex);
+        if(unlock_ret != 0){
+            syslog(LOG_ERR, "Error occured during mutex lock = %s. Exiting...", strerror(errno));
+            exit(-1);  
+        }
 
-#define LOGFILE_PATH	("/var/tmp/aesdsocketdata")
-#define BUFFER_SIZE	(1024) //chosen as buffer size, can be reduced to 512 as well
+        //Get the timestamp
+        char timestamp[256];
+        time_t curr_time;
+        struct tm *t2;
 
-//Global to handle signal exit
-bool graceful_exit_handler = false;
+        curr_time = time(NULL);
+        t2 = localtime(&curr_time);
+        if (t2 == NULL){
+            syslog(LOG_ERR, "Error occured during localtime(). Exiting...");
+            goto mutex_release; 
+        }
+        strcpy(timestamp, "timestamp:");
+        strftime(&timestamp[10], sizeof(timestamp) - 10, "%a, %d %b %Y %T %z", t2);
+        char *timestamp_buffer = malloc(sizeof(char) * strlen(timestamp) + 2);
+        if(!timestamp_buffer){
+            syslog(LOG_ERR, "Error occured during malloc() for timestamp = %s. Exiting...", strerror(errno));
+            goto mutex_release; 
+        }
+        sprintf(timestamp_buffer, "%s%s", timestamp, "\n");
+
+        if(lseek(file_fd, 0, SEEK_END) == -1){
+            syslog(LOG_ERR, "Error occured during lseek() to EOF = %s. Exiting...", strerror(errno));
+            goto exit_label1; 
+        }
+        
+        int written_bytes;
+        int len_to_write = strlen(timestamp_buffer);
+        char *ptr_to_write = timestamp_buffer;
+        while(len_to_write != 0){
+            written_bytes = write(file_fd, ptr_to_write, len_to_write);
+            if(written_bytes == -1){
+                if(errno == EINTR)
+                    continue;
+
+                syslog(LOG_ERR, "Error occured during write() = %s. Exiting...", strerror(errno));
+                goto exit_label1; 
+            }
+            len_to_write -= written_bytes;
+            ptr_to_write += written_bytes; 
+        }
+
+        file_size += strlen(timestamp_buffer);
```

Comment:
@prvbharadwaj1731 it's going to be a challenge to make the accesses to `file_size` all be atomic and you are going to end up with a mutex that locks a lot of different things.  I don't think you necessarily need to do this, think about a design where this isn't necessary.